src/coprocessor/codec/mysql/time/tz.rs

```diff
@@ -56,6 +66,17 @@ impl Tz {
 
     /// Constructs a local time zone.
     pub fn local() -> Self {
+        // Attempt to resolve a time zone name first, because somehow `chrono::Local` is slow.
+        if let Some(local_time_zone_name) = LOCAL_TIME_ZONE_NAME.as_ref() {
```

Comment:
In what conditions that local timezone will be used? What if TiKV nodes are located all over the world, are the query results still the same?

Reply:
When TiDB explicitly pushes the SYSTEM time zone. There will be issues in such scenario and currently this is regarded as an intended behaviour.

Reply:
Why is intended? I think the proper behavior should be pushing down the local timezone of TiDB instance or client instead of relying on the TiKV themselves.

Reply:
Golang cannot retrieve current time zone as well. @BusyJay

Reply:
That doesn't justify relying on TiKV local timezone. For retrieving time zone in golang, does it work for you: https://play.golang.org/p/xmKMiASgMbr?

Reply:
It won't work, for example, CST represents for Central Standard Time, Central America Standard Time, Cuba Standard Time, Canada Central Standard Time, China Standard Time and Taipei Standard Time. In Golang, inferring actual time zone name is a TODO item. I and @zhexuany has already discussed about this and we decided to push down "local" to TiKV currently.

Reply:
The guessing trick doing in TiKV should also applies to TiDB. Pushing down "local" is wrong.

Reply:
Yes. Pushing down "local" is wrong, hoping one day TiDB can fix it, pushing down the actual time zone name instead of letting TiKV judge. However currently TiDB cannot provide such functionality and evaluation local is the only workaround way available, so let's simply fix the performance issue in this local first.

Reply:
Summary:

1. Yes. Pushing down local is wrong.
2. Not pushing down local requires hard work, which cannot be done in short time. As a *workaround*, local is pushed down to TiKV instead.
3. This PR is trying to fix the performance issue of this workaround.

Reply:
I suggest to fix it in the TiDB side. TiDB can detect what the local timezone is when startup. If it fails to get an accurate timezone id, then fallback to SYSTEM or offset base solution or even rejects to start. It has similar improvement comparing with this PR and also is the *correct* way to start fixing the problem.

Reply:
I agree. However this issue is currently a release blocker for rc2. How about postpone this task to rc3 and let's fix it in TiKV side in rc2?